web/interceptor/access_control.go

```diff
@@ -0,0 +1,157 @@
+	"CreateGroup":             {group, teacher},
+	"GetGroup":                {group, teacher},
+	"GetAssignments":          {student, teacher},
+	"GetEnrollmentsByCourse":  {student, teacher},
+	"GetRepositories":         {student, teacher},
+	"UpdateGroup":             {teacher},
+	"DeleteGroup":             {teacher},
+	"GetGroupsByCourse":       {teacher},
+	"UpdateCourse":            {teacher},
+	"UpdateEnrollments":       {teacher},
+	"UpdateAssignments":       {teacher},
+	"UpdateSubmission":        {teacher},
+	"UpdateSubmissions":       {teacher},
+	"RebuildSubmissions":      {teacher},
+	"CreateBenchmark":         {teacher},
+	"UpdateBenchmark":         {teacher},
+	"DeleteBenchmark":         {teacher},
+	"CreateCriterion":         {teacher},
+	"UpdateCriterion":         {teacher},
+	"DeleteCriterion":         {teacher},
+	"CreateReview":            {teacher},
+	"UpdateReview":            {teacher},
+	"GetReviewers":            {teacher},
+	"IsEmptyRepo":             {teacher},
+	"GetSubmissionsByCourse":  {teacher, courseAdmin},
+	"GetUserByCourse":         {teacher, admin},
+	"GetUsers":                {admin},
+	"GetOrganization":         {admin},
+	"CreateCourse":            {admin},
+}
+
+func AccessControl(logger *zap.SugaredLogger, tm *auth.TokenManager) grpc.UnaryServerInterceptor {
+	return func(ctx context.Context, req interface{}, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (interface{}, error) {
+		method := info.FullMethod[strings.LastIndex(info.FullMethod, "/")+1:]
+		logger.Debugf("ACCESS CONTROL for method %s", method) // tmp
+		roles, ok := access[method]
+		if ok {
+			logger.Debug("Got roles: ", roles) // tmp
+			claims, err := tm.GetClaims(ctx)
+			if err != nil {
+				logger.Error("Access control: failed to get claims from request context: %v", err)
+				return handler(ctx, req)
+			}
+			logger.Debug("Got user claims: ", claims) // tmp
+			for _, role := range roles {
+				switch role {
+				case none:
+					return handler(ctx, req)
+				case user:
+					if m, ok := req.(requestID); ok {
+						if m.IDFor("user") == claims.UserID {
+							log.Printf("IDs match:%d, %d", m.IDFor("user"), claims.UserID)
+							return handler(ctx, req)
+						} else {
+							log.Printf("IDs don't match:%d, %d", m.IDFor("user"), claims.UserID)
+						}
+					}
+				case group:
+					if m, ok := req.(requestID); ok {
+						groupID := m.IDFor("group")
+						for _, group := range claims.Groups {
+							if group == groupID {
+								return handler(ctx, req)
+							}
+						}
+					}
+				case student:
+					if m, ok := req.(requestID); ok {
+						courseID := m.IDFor("course")
+						status, ok := claims.Courses[courseID]
+						if ok && status == qf.Enrollment_STUDENT {
+							return handler(ctx, req)
+						}
+					}
+				case teacher:
+					// TODO(vera): needs different handling for "GetUserByCourse"
+					if m, ok := req.(requestID); ok {
+						courseID := m.IDFor("course")
```

Comment:
I just noticed that `"course"` is not a role. Hence, it seems like the wrong idea to make the interface be:
```go
type requestID interface {
	IDFor(qf.Role) uint64
}
```
As I proposed above... (need to reconsider my recommendation here...)